### PR TITLE
Add field reporter to trait Verifier - adapting Carbon

### PR DESCRIPTION
**Pull request by aterga**
Should be merged after [Silver PR 513](https://github.com/viperproject/silver/pull/513).

### CODE REVIEW

**Comment by aterga** [comment]
The new functionality is that ```BoogieInterface``` now reports ```BackendSubProcessReport``` messages to indicate at which communication stage with a Boogie process it's in. This is needed to make sure the IDE does not leak Z3 processes while used with Carbon. The problem is that Boogie sometimes terminates successfully while leaving its Z3 sub-process running. This occurs e.g. if there are matching loops in the SMT query. By the time the IDE decides to terminate a verification process (e.g. after a timeout) we can no longer find the PID of the leaked Z3 process as the parent Boogie process is already gone. We could collect the PIDs in advance, but for that we need to either poll (which is suboptimal) tr to know the moment in time when Carbon has already started a Boogie process. This is an example situation in which the client benefits from such reports.

File: src/main/scala/viper/carbon/CarbonVerifier.scala [hunk 1]
```diff
@@ -13,17 +13,18 @@ import viper.silver.utility.Paths
 import viper.silver.verifier._
 import verifier.{BoogieDependency, BoogieInterface, Verifier}
 import java.io.{BufferedOutputStream, File, FileOutputStream, IOException}
+
```
Comment (gauravpartha) [inline src/main/scala/viper/carbon/CarbonVerifier.scala:16]: remove line

File: src/main/scala/viper/carbon/CarbonVerifier.scala [hunk 2]
```diff
@@ -13,17 +13,18 @@ import viper.silver.utility.Paths
...
 
 /**
  * The main class to perform verification of Viper programs.  Deals with command-line arguments, configuration
  * of modules and choosing which module implementations to use.
  *
  * Debug information can either be set using the constructor argument or the setter.
  */
-case class CarbonVerifier(private var _debugInfo: Seq[(String, Any)] = Nil) extends Verifier with viper.silver.verifier.Verifier with BoogieInterface {
-
-  def this() = this(Nil)
```
Comment (gauravpartha) [inline src/main/scala/viper/carbon/CarbonVerifier.scala:26]: I guess def this() is never used?

File: src/main/scala/viper/carbon/verifier/BoogieInterface.scala [hunk 3]
```diff
@@ -138,10 +143,12 @@ trait BoogieInterface {
     var res: String = ""
     var reserr: String = ""
     def out(input: java.io.InputStream): Unit = {
+      reporter report BackendSubProcessReport("carbon", boogiePath, onOutput/*, _boogieProcessPid*/)
```
Comment (gauravpartha) [inline src/main/scala/viper/carbon/verifier/BoogieInterface.scala:146]: It's a bit strange that "onOutput" is lower case, while the other stages are upper case (such as "OnError"). But this is related to https://github.com/viperproject/silver/pull/513.

**Review by gauravpartha — CHANGES_REQUESTED** [review]

**Comment by gauravpartha** [comment]
I guess the uncommented arguments ` _boogieProcessPid` are for https://github.com/viperproject/carbon/pull/384?

File: src/main/scala/viper/carbon/CarbonVerifier.scala (hunk 2 above)
Comment (aterga) [reply src/main/scala/viper/carbon/CarbonVerifier.scala:26]: Right, that's why I removed it.

**Comment by aterga** [comment]
> I guess the uncommented arguments ` _boogieProcessPid` are for #384?

Exactly. 

Thanks for your review!

**Comment by aterga** [comment]
No need to merge this. This PR has been incorporated into PR#384

**Pull request closed without merging**